### ISSUE TITLE
shortcut to make the editor invisible

### ISSUE THREAD

**Issue by ralf3u** [xiota/geanylua-scripts#4]
it could be that sometimes one will see only the content of the preview

possible solution
in Preferences, in Keybindings: toggle editor on/off

**Comment by xiota** [comment]
Would a script that resizes the sidebar do what you want?  You could set a list of percentages, like 30%, 70%, 100%.  Then it would cycle through them.  That might avoid the current problems with toggle-editor.

**Comment by ralf3u** [comment]
>>3
If the vertical line between editor and preview is at about one third at the right side, and then if I press many times the shortcut Toggle editor, sometimes a vertical line appears just for a part of a second at about one third of the window on the left side.
>
>Would a script that resizes the sidebar do what you want? You could set a list of percentages, like 30%, 70%, 100%. Then it would cycle through them. That might avoid the current problems with toggle-editor.

I think that there is a misunderstanding. I just recognized that not all html-documents show a vertical line for a part of a second. It seems to be that some html-documents are shown sometimes with an artifact just for a part of a second, is it the vertical line or something else. Maybe it is because the used computer is slow. Down below is an example of html-code. In the moment where I press the shortcut "Toggle editor", I can see sometimes not only the green color in the middle, but also for a part of a second a green rectangle on the left side, specially when clicking fast many times  the shortcut "Toggle editor". Do you see it too?

```
<!DOCTYPE html>
<html>
<head>
<title>test preview</title>
<style type="text/css">
.content {position: absolute;left: 50%;top: 50%;transform: translate(-50%, -50%)}
</style>
<body>
<div class="content" style="width:830px;height:590px;background:green"></div>
</body>
</html>
```

**Comment by xiota** [comment]
If I hold down the shortcut and let it run as fast as it can, I see green on the left (normally my sidebar is on the right).  It's probably a GUI update/refresh that I can't do anything about.  But, would resizing the sidebar instead of hiding the editor be a better solution (if possible in Lua)?

**Comment by ralf3u** [comment]
>Unfortunately, I don't see anything that can be used to resize the sidebar.

I didn't ask to resize the sidebar. The vertical line that appears just for a part of a second somewhere else in Geany, appears not in all html-documents, but only in some html-documents. And it seems to be that the GUI update/refresh is the cause that creates some artifacts for a part of a second, like the green color mentioned in your last comment. 

To 4 and 5:
Can it be that it is too complicated or not possible to program, because then it would need also to re-code the shortcut "Toggle Sidebar" (specially 5)? Or is it because of the argument that coding crisscross things can cause a lot of problems? Or is it because you don't think that it is useful that shortcuts do crisscross things (what I can understand)?

focus feature
a
If the editor is in focus and if the shortcut "Toggle editor" is pressed, then sidebar has full width, but the content of the Preview-tab is not in focus. Better would be that if the Preview-tab is active, then the content of the Preview-tab should be automatically in focus.
b
When leaving sidebar full width with the shortcut "Toggle editor", the editor is not in focus. Better would be when editor is in focus automatically, because in the most cases before going in sidebar full width, the editor is in focus.

Would it be possible to add the two focus features a and b to the shortcut "Toggle editor"?

**Comment by xiota** [comment]
> I didn't ask to resize the sidebar.

I know you didn't ask for that, but doing so would have provided the feature you want (maximizing the sidebar) without the other problems.

> the GUI update/refresh is the cause

I think so too.  It is from GTK and out of my control.  To fix it (in GTK) would probably require low level changes that would cause more problems.  Since it's only momentary, it's better to leave it alone.

> To 4 and 5:

The toggle scripts could look for the state files of each other's toggle scripts.  It makes the code complicated, and it's better not to if it isn't needed.

> focus feature

The GeanyLua API is limited, so it might not be possible.  I'll check.

**Comment by ralf3u** [comment]
>The GeanyLua API is limited, so it might not be possible. I'll check.

That would be great. Maybe this link https://github.com/xiota/geany-preview/issues/4#issuecomment-961720440 could help, because also at the beginning of the development of the shortcut "Toggle between the editor and preview", the content of the Preview-tab was not in focus.

**Comment by xiota** [comment]
> focus feature

It might work now.  Let me know.

**Comment by ralf3u** [comment]
>It might work now. Let me know.

Now when toggling, the tab-title of the Preview is in focus and then the tab-title of the editor is in focus.
The goal was that the content of the Preview is in focus and then the content of the editor is in focus, so in both cases not the tab-title.
I think that I didn't explain it correctly in my other comment when I meant the editor should be in focus. With that, I was meaning that the highlighted line should appear again in the editor (on that place where it was the last time to continue coding).

In the end the goal is that the focus of the shortcut "Toggle editor" should work exactly like the focus of the shortcut "Toggle between the editor and preview", in the editor and in the preview.

**Comment by xiota** [comment]
> the tab-title of the Preview is in focus and then the tab-title of the editor is in focus

Lua doesn't have any way to get the contents of the ta

**Comment by xiota** [comment]
I've reverted the change.

**Comment by ralf3u** [comment]
>Trying to hiding gui elements and grab focus hasn't worked well in the past (eg, toggle focus between editor and sidebar).

I don't understand. The shortcut "Toggle focus between the editor and sidebar" works absolutely perfect. There is not even one thing that does not work. There is no issue open about that.

**Comment by xiota** [comment]
> I don't understand. The shortcut "Toggle focus between the editor and sidebar" works absolutely perfect. There is not even one thing that does not work. There is no issue open about that.

Maybe I'm mis-remembering.  I thought there were problems with tab navigation or something.  Maybe I'm confusing it with something else.

GeanyLua references widgets by their name in the glade file.  I don't think it's possible to access GTK widgets that are created at runtime, like the contents of specific tabs.  There's another scripting plugin ([peasy](https://github.com/kugel-/peasy/)) that *might* be able to do it.  I haven't tried writing any scripts for it yet.

**Comment by ralf3u** [comment]
>Maybe I'm mis-remembering. I thought there were problems with tab navigation or something.

There was a problem with tab-browsing (https://github.com/xiota/geany-preview/issues/5#issuecomment-939473728; in the comment it is called "issue 3"). It was fixed (https://github.com/xiota/geany-preview/issues/5#issuecomment-956119413), then reappeared (https://github.com/xiota/geany-preview/issues/11), and now just disappeared (https://github.com/xiota/geany-preview/issues/11#issuecomment-1155150159). It was only visible when the Symbols-tab was part of the sidebar (https://github.com/xiota/geany-preview/issues/11#issuecomment-1039649263), so it was just a very small issue.

Tomorrow I would like to open in Geany Preview a new issue about the shortcut "Toggle sidebar". Then the next day I would like to open again an issue, but this time about the shortcut "Toggle editor". If you don't like both reports, just delete them, and copy those parts of text that you think could be interesting, maybe to paste the parts here in this issue.

**Comment by ralf3u** [comment]
>I've reverted the change.

Thank you so much for the shortcut "Toggle editor". It is a significant option, because now there's no need to switch to the browser to see the webpage in big, if one don't want to resize the sidebar with the mouse.

**Comment by ralf3u** [comment]
It's quite interesting, because in the second comment of this issue I noted:

>This feature could be specially interesting for webdesign on small screens. I think that this would be a cool feature.

To that time, I didn't use  a small display. I even didn't think to use a small display. But things have changed, and now I'm using a small display. Well, it is not really small, but the width is not so big. So, the shortcut to make the editor invisible is for me very interesting.
Because all focus issues of the plugin Preview are now fixed on your computer, does also this shortcut set the focus correctly on your computer?

The best would be if the shortcut works like this:
Toggle between only-sidebar-view (focus is on the content of the Preview-tab) and only-editor-view (focus is on the content of the editor).

**Comment by xiota** [comment]
With the `switch-views` scripts, focus goes where I expect.  (Geany 2.0, GTK 3.24.41)  When only editor is visible, focus goes to editor.  When only sidebar is visible, focus goes to contents of sidebar.  But if tab is changed by user, focus stays on tabbar, unless xitweaks plugin is loaded with focus highlighting feature enabled.  Then focus goes to tab contents (by xitweaks).

I've updated the scripts to attempt to change focus at the same time editor/sidebar visibility is changed.  I think before I tried to add some delay to work around some issue, but it never really worked and now no longer needed.
